### PR TITLE
H374 — Self-Consistency vs Frozen EP13 EMA Teacher on τ_z Channel (λ=0.1)

### PR DESCRIPTION
**Pull request by morganmcg1**
## Hypothesis

**Add τ_z-only self-consistency regularization vs frozen EP13 EMA teacher during EP14-EP16 cosine tail. Anchors fine-tuning trajectory to the EP13 manifold on the noisiest channel.**

Mechanism: Freeze the EP13 EMA checkpoint as a teacher. During fine-tuning, add a consistency loss:
```
L_consistency = λ × MSE(student_τ_z, teacher_τ_z.detach())
```
applied **only on the τ_z channel**, with λ=0.1. Teacher gradients are detached; teacher is frozen (no grad updates). This is **zero added parameters** — the consistency loss is computed against the cached EP13 teacher prediction.

**Why this attacks WSS_z specifically:**
- test_WSS_z = 8.61% — 2.4× larger than WSS_x (5.35%) and 2.4× larger than the AB-UPT target (3.63%). All other channels (WSS_x, WSS_y, p_s, p_v) are at or below target.
- The CFD labels for τ_z are the **noisiest** (most sensitive to mesh resolution and solver convergence). Empirically, the 21-axis closure log shows the EP13 → EP15 fine-tune trajectory drifts on τ_z in ways the loss signal cannot correct.
- The EP13 EMA has already learned a reasonable τ_z manifold (the basis of all SOTA recipes since H244). Fine-tuning EP14-EP16 without an anchor allows over-fitting to noisy τ_z samples while WSS_x/y are well-behaved.
- Consistency anchors the τ_z manifold drift WITHOUT preventing improvement on other channels. The teacher provides a "soft prior" only where the student is uncertain.

**Distinct from prior nulls:**
- Loss reweighting (7 nulls): scalar reweighting of τ_z gradient magnitude — closed. **This is a NEW LOSS TERM, not a reweighting of existing loss.**
- TTA / cal (15+ nulls): inference-time only. **This changes the TRAINING manifold.**
- Encoder / decoder / input-feature axes (13 nulls combined): structural changes to forward pass. **This is purely a training-time regularizer — zero structural change.**
- Mean Teacher (Tarvainen & Valpola, NeurIPS 2017) and NoRD (Wang et al., NeurIPS 2024) — consistency vs frozen EMA reduces error on noisy-label regimes (CFD τ_z is the analog).

**Edward — your 5/5 close-rule streak (H338/H361/H364/H368/H370/H371) has bounded the loss-reweighting, encoder, and operator-replacement tiers cleanly. This hypothesis exits those tiers entirely: it's a training-time regularizer with zero structural change. Different axis, fresh look.**

## Instructions

**Branch:** `edward/h374-self-consistency-ep13-teacher-tauz` (will create)

**Step 1 — Add CLI flags to `target/train.py`:**
```
--teacher-ckpt "/path/to/ep13_ema.pt"   # frozen EP13 EMA as teacher
--self-consistency-weight 0.1            # λ for consistency loss (start 0.1)
--self-consistency-channels wss_z        # apply ONLY to τ_z channel
--self-consistency-loss mse              # L2 consistency
```

Implementation:
- Load teacher checkpoint at init, set `requires_grad=False`, set `.eval()` mode.
- In training step: after the student forward pass and student loss computation:
  - Run teacher forward pass on the SAME input batch (gradient detached, no_grad context).
  - Extract teacher's τ_z predictions (channel 4 of surface_y, after model output — match the same normalization the student uses).
  - Compute `L_consistency = λ × MSE(student_τ_z_pred, teacher_τ_z_pred.detach())`.
  - Total loss: `L_total = L_student + L_consistency`.
- Teacher EVAL mode is fixed across training; do not update teacher.
- Important: extract ONLY the τ_z slice from the prediction tensor (channel index where τ_z lives in surface_y). DO NOT apply consistency to τ_x, τ_y, p_s, or volume.

**Step 2 — Smoke check (1-epoch debug, batch_size=4, no warm-start):**
```bash
torchrun --standalone --nproc-per-node=8 train.py \
  --epochs 1 --debug \
  --teacher-ckpt "outputs/ensemble_cache/run-yw2a5dyl-epoch-13-ema/checkpoint.pt" \
  --self-consistency-weight 0.1 --self-consistency-channels wss_z --self-consistency-loss mse \
  --wandb-group h374-edward-smoke \
  --wandb-name "edward/h374-smoke-consistency"
```
Verify: loss decomposition logged (student_loss, consistency_loss separate scalars), no NaN, training loss decreasing.

**Step 3 — Phase 1 (EP13 → EP14-EP16 from EP13 EMA, 8 GPUs DDP, 3-epoch cosine tail):**
```bash
H185_EP13_EMA="outputs/ensemble_cache/run-yw2a5dyl-epoch-13-ema/checkpoint.pt"

torchrun --standalone --nproc-per-node=8 train.py \
  --warm-start-from "$H185_EP13_EMA" \
  --teacher-ckpt "$H185_EP13_EMA" \
  --self-consistency-weight 0.1 --self-consistency-channels wss_z --self-consistency-loss mse \
  --epochs 3 --start-epoch 13 \
  --optimizer lion --lr 9e-5 --lr-cosine-t-max 3 --lr-min 1e-6 \
  --weight-decay 5e-4 --batch-size 4 \
  --tau-y-loss-weight 1.3 --tau-z-loss-weight 1.67 \
  --surface-loss-weight 2.0 --volume-loss-weight 0.5 \
  --grad-clip-norm 0.5 --ema-decay 0.999 \
  --use-qk-norm --rff-num-features 16 --pos-encoding-mode string_separable \
  --model-layers 5 --hidden-dim 512 --heads 4 --slices 128 \
  --vol-points-schedule "0:65536" \
  --wandb-group h374-edward-self-consistency \
  --wandb-name "edward/h374-phase1-lambda0p1"
```

**Pre-committed kill rules:**
- EP14 val_primary > 6.00% → CLOSE immediately
- EP14 val_WSS_z > H336 EP13 baseline (~9.06%) → CLOSE (consistency over-anchors)
- EP14 val_primary ≤ 6.00% AND val_WSS_z drops ≥30bp → continue, run TTA cascade
- If Phase 1 lands marginal (val_primary 6.00-6.05%), launch a fast Phase 2 with λ=0.05 on a 1-epoch reseed — DO NOT escalate to λ=0.2+ (over-regularization risk)

**Step 4 — TTA cascade (if Phase 1 passes):**
Apply H336 K=5 + Student-t ν=4 + 8-res + mirror TTA + per-channel cal on best ckpt, then test split.

## Baseline

**Current SOTA H342 (PR #1526, MERGED):**
- val_abupt_calibrated = **5.8962%**
- test_abupt_calibrated = **5.7357%**
- test_VP = 3.3751%, test_SP = 3.6124%, test_WSS = 6.6351%, **test_WSS_z = 8.6122%**

**Merge gate (strict AND):** val_cal < 5.8962% AND test_cal < 5.7357%

**Per-channel error budget (τ_z is the bottleneck):**
| Channel | H342 test | AB-UPT target | Gap |
|---------|-----------|---------------|-----|
| **WSS_z** | **8.6122%** | 3.63% | **4.98pp** |
| WSS_x | 5.3512% | 5.35% | 0.00pp |
| WSS_y | 3.6488% | 3.65% | 0.00pp |
| p_s | 3.8201% | 3.82% | 0.00pp |
| p_v | 3.3735% | 3.42% | 0.00pp |

**Estimated wall-clock:** ~3.5h/epoch × 3 epochs = ~10.5h on 8 GPUs (Phase 1, slightly slower due to teacher forward pass). TTA cascade ~7h. Total ≤ 18h.

**References:**
1. NoRD (Wang et al., NeurIPS 2024) — "Noise-Resilient Self-Distillation"; consistency vs frozen EMA reduces error on noisy-label benchmarks.
2. Mean Teacher (Tarvainen & Valpola, NeurIPS 2017, arxiv:1703.01780) — foundational EMA-teacher consistency reference.
3. Kaltenbach & Koutsourelakis (2021) — "Knowledge Distillation for Physical Simulations" — distillation from fixed checkpoint improves OOD generalization on wall-bounded flows.

**Constraints:**
- DDP×8 mandatory
- τ_z normalization LOCKED at 1.67 (never modify)
- Read-only: `data/loader.py`, `data/preload.py`, `data/split_manifest.json`
- Zero parameters added (teacher frozen, only λ scalar)
- No ensembles

Use `--wandb_group h374-edward-self-consistency`. Submit terminal `SENPAI-RESULT` marker on completion.

**Kill v1 H371 ranks first if still running** (PR #1566 closed Finding V; the rank0/1/2/...rank7 runs in W&B `edward/h371-isab-m32-1layer-idx2-phase1-rank*` are wasted GPU). After cleaning up, launch H374 fresh.

### CODE REVIEW

**Comment by morganmcg1** [comment]
STUDENT edward: Picked up. EP13 EMA teacher checkpoint is at `outputs/drivaerml/resume_cache/yw2a5dyl/epoch-13/checkpoint.pt` (not the path the PR shows; `outputs/ensemble_cache/run-yw2a5dyl-epoch-13-ema/checkpoint.pt` does not exist on this pod). The cached file is from run yw2a5dyl epoch 13 with `checkpoint_source=ema` — that's the right one. Proceeding with this path. Will research the consistency-regularization choice briefly, then smoke + Phase 1.

**Comment by morganmcg1** [comment]
STUDENT edward: Smoke + Phase 1 update.

**Smoke (in-process, single GPU):** PASS. Teacher loads with 0 missing/0 unexpected keys, consistency_loss==0 when student.eval()==teacher.eval(), grows with perturbation (0.145 at student-perturbation σ=1e-2), gradient flows through both backbone and surface_out head, teacher gradients stay None. Test at `tools/h374_smoke_test.py`.

**Phase 1 launched:**
- W&B run: `98kd6tt5` (rank 0). Group `h374-edward-self-consistency`. Other ranks: 89pohoo0, ...
- Log: `logs/h374/phase1_20260602_131339Z.log`
- Teacher: `outputs/drivaerml/resume_cache/yw2a5dyl/epoch-13/checkpoint.pt` (EP13 EMA, the same file the PR cited; just under a different relative path on this pod)
- Recipe: lr=9e-5, T_max=3 cosine (fresh tail, EP14 lr=6.775e-5 → EP16 ~1e-6), epochs_already_done=13, λ=0.1 MSE on τ_z channel only, all other H342/H336 recipe-locked flags carried (mirror aug, drop_path 0.1, full-65k vol-points, etc.)

**Open questions (will pause if you respond):**
- The PR command shows `--warm-start-from` and `--start-epoch 13` which aren't existing flags on tay. I used the canonical `--resume-from-wandb yw2a5dyl --resume-alias epoch-13 --epochs-already-done 13` mechanism plus `--epochs 16`, consistent with H338/H361. Please flag if that's wrong.
- The PR specified `--lr-cosine-t-max 3`, which after the 13-step scheduler advance puts EP14 at lr≈6.775e-5 (75% of peak). This is the 'fresh aggressive tail' recipe used in H361, not the H338-style `--lr-cosine-t-max 16` tail (lr≈8.5e-6). I went with what the PR says.

Wakeups scheduled to monitor EP14 kill rules (val_primary > 6.00% OR val_WSS_z > 9.06% → CLOSE).

**Comment by morganmcg1** [comment]
STUDENT edward:
SENPAI-RESULT: {"terminal":true,"status":"closed","pending_arms":false,"wandb_run_ids":["98kd6tt5","89pohoo0","sohqk3nt","b5tb0v8x","2tcvh6br","fus1uoab","i6arnk3g","oim8xb3y"],"primary_metric":{"name":"val_primary_uncal_ep14","value":6.0700},"test_metric":{"name":"test_abupt_calibrated","value":null}}

## Results — KILLED on EP14 (both pre-committed kill rules fired)

**EP14 validation (uncalibrated, raw val_primary; the only completed validation before kill):**

| Metric | H374 EP14 | H342 SOTA (val_cal) | Δ | Kill rule |
|---|---|---|---|---|
| **val_primary** | **6.0700%** | 5.8962% | +17bp | **>6.00% → CLOSE** ❌ |
| **val_WSS_z** | **9.2573%** | (H336 EP13 ~9.06%) | +20bp | **>9.06% → CLOSE** ❌ |
| val_WSS_x | 5.9697% | 5.35% | +62bp | — |
| val_WSS_y | 7.5308% | 3.65% | +388bp | — |
| val_WSS_total | 6.8582% | — | — | — |
| val_surface_pressure | 4.0069% | 3.82% | +19bp | — |
| val_volume_pressure | 3.5853% | 3.42% | +17bp | — |

Both pre-committed kill rules fired simultaneously after the first cosine-tail epoch. 6.07% is above the 6.00-6.05% marginal-reseed range, so per the PR instructions I did **NOT** launch the Phase 2 λ=0.05 reseed and did **NOT** run the TTA cascade. Phase 1 was killed at EP15 21% (~1h22m wall-clock), saving ~2h of GPU time.

## Consistency loss diagnostics (sanity)

- Teacher load: 0 missing, 0 unexpected keys (verified in smoke and confirmed in run init)
- Smoke test: `tools/h374_smoke_test.py` — teacher.eval()==student.eval() → consistency==0; perturbation σ=1e-2 → 0.145; teacher.grad stays None ✓
- Training-step consistency_loss (raw L2 on τ_z): mean=0.00162, max=0.0297, first=0.000226, last=0.000571
- Weighted (λ=0.1 × raw): mean=0.000162

So the consistency term *was* active, the τ_z channel slice was correctly extracted, and gradients flowed. **The hypothesis was tested cleanly — it just didn't work.**

## Command used

```bash
torchrun --standalone --nproc-per-node=8 train.py \
  --resume-from-wandb yw2a5dyl --resume-alias epoch-13 --epochs-already-done 13 \
  --teacher-ckpt outputs/drivaerml/resume_cache/yw2a5dyl/epoch-13/checkpoint.pt \
  --self-consistency-weight 0.1 --self-consistency-channels wss_z --self-consistency-loss mse \
  --epochs 16 \
  --optimizer lion --lr 9e-5 --lr-cosine-t-max 3 --lr-min 1e-6 \
  --weight-decay 5e-4 --batch-size 4 \
  --tau-y-loss-weight 1.3 --tau-z-loss-weight 1.67 \
  --surface-loss-weight 2.0 --volume-loss-weight 0.5 \
  --grad-clip-norm 0.5 --ema-decay 0.999 \
  --use-qk-norm --rff-num-features 16 --pos-encoding-mode string_separable \
  --model-layers 5 --hidden-dim 512 --heads 4 --slices 128 \
  --vol-points-schedule "0:65536" \
  --wandb-group h374-edward-self-consistency \
  --wandb-name "edward/h374-phase1-lambda0p1"
```

- **Wall-clock to kill:** 1h22m (EP14 done, EP15 at 21%)
- **Peak GPU memory:** ~80 GB reserved / GPU (val/pre_clear_reserved_gb=79.7 GB; comparable to H342)
- **EP14 train epoch_loss:** 0.00893 (consistency-augmented total)
- **Run IDs (DDP×8):** 98kd6tt5 (rank0), 89pohoo0, sohqk3nt, b5tb0v8x, 2tcvh6br, fus1uoab, i6arnk3g, oim8xb3y
- **Group:** `h374-edward-self-consistency`

## What happened — honest analysis

The hypothesis predicted that anchoring τ_z to a frozen EP13 EMA teacher would *reduce* WSS_z drift during fine-tuning. The opposite happened:

1. **WSS_z got worse, not better.** EP14 val_WSS_z = 9.26%, above the H336 EP13 baseline (~9.06%) — the second kill rule. The teacher's frozen EP13 manifold is *itself* at ~9% on the noisy τ_z channel; anchoring to it cannot improve on what the teacher already has, and the consistency penalty actively pulled the student back toward that suboptimal manifold while the student's own loss was trying to escape it.
2. **The other channels regressed mildly too** (WSS_x +62bp, WSS_y +388bp vs the H342 calibrated targets), suggesting the consistency penalty siphoned some gradient capacity away from the fine-tune signal even on channels we didn't constrain — likely via the shared backbone, since the τ_z consistency gradient flows through the entire encoder before fanning out at the heads.
3. **The "noisy CFD label" framing was wrong for this teacher.** The Mean Teacher / NoRD prior assumes the teacher's predictions are *closer to the clean target* than the student's noisy supervision. Here the teacher is the *same model family at the same fine-tune stage* — it has the same WSS_z error pattern as a fresh fine-tune would. Consistency to it is just a soft L2 prior toward an arbitrary frozen point in the same loss basin, not a denoising signal.

## Suggested follow-ups

Filing as a clean null. The closure information: **τ_z consistency vs frozen-EP13-EMA teacher with λ=0.1 over-anchors** — confirming the H336 reweight family's lesson that simply pulling τ_z back toward a known-suboptimal manifold does not work, regardless of whether the pull is via a static loss weight (H336/H361/H364 family) or a soft consistency target (H374). The teacher's information content on τ_z is not better than the student's.

Possible adjacent directions (NOT implementing here, just observations):
- **A different teacher.** Consistency-vs-self-frozen is closed; consistency vs a *higher-quality* τ_z signal (e.g. a τ_z-only model trained with a different loss / different aug, or an ensemble-averaged teacher) might still have headroom.
- **Channel-conditional consistency.** Apply consistency only on points where the student is high-confidence and teacher disagrees — selectively, not uniformly. This is much more expensive but might avoid the "drag toward suboptimal" failure mode.
- **Different target.** Consistency on intermediate representations (slice-token features, encoder outputs) rather than the τ_z head output — possibly more useful as a representational regularizer than as a prediction-level prior.

But I would explicitly **not** recommend another λ-sweep of the same recipe. The kill rules diagnosed the failure mode (consistency over-anchors a same-family teacher), not a hyperparameter setting issue.

**Comment by morganmcg1** [comment]
ADVISOR: Closing PR #1569 because ADVISOR: Closing H374 as a clean null. Both pre-committed kill rules fired at EP14: val_primary=6.07% (>6.00% threshold) and val_WSS_z=9.26% (>9.06% baseline). 

Edward's analysis is correct and well-reasoned. The core issue: the EP13 EMA teacher itself has ~9% WSS_z error, so anchoring τ_z to it cannot improve beyond the teacher's own ceiling. The consistency penalty actively pulled against the fine-tune signal, and the shared backbone propagated the τ_z gradient suppression to the other channels (WSS_y +388bp is the clearest symptom).

This cleanly falsifies the τ_z-vs-frozen-EP13-EMA consistency regularization approach regardless of λ setting. The failure mode is not hyperparameter sensitivity — it's that the teacher's τ_z manifold is not higher quality than the student's.

Finding W already assigned (H366 hierarchical kNN, PR #1560). This is a separate, clean null for the TRAINING REGULARIZATION tier on τ_z.

Edward now idle — new assignment incoming..
